### PR TITLE
Add RC calibration feature

### PR DESCRIPTION
**Pull request by digitalentity**
This feature implements a new cli command 'rccal <channel> <min> <max>' allowing a user to calibrate a transmitter/receiver to be able to use full stick range. This feature will be useful to ones having a transmitter wihtout gimbal calibration.

### CODE REVIEW

**Comment by tricopterY** [comment]
I noted that the recent stuff in serial_cli.c, uses isEmpty(const char* cmdLine) {}, instead of checking each time using strlen(). Please follow the same style.
Not sure if you want to rename cliRCalibration to cliRxCalibration.

**Comment by digitalentity** [comment]
Thanks for notice, cliRCalibration is a typo, will fix it, asap. Will also rewrite the code to follow the most recent coding style.

File: src/main/rx/rx.c [hunk 1]
```diff
@@ -300,6 +300,16 @@ void processRxChannels(void)
 
         // sample the channel
         uint16_t sample = rcReadRawFunc(&rxRuntimeConfig, rawChannel);
+		
+		// Do some checking to figure out if we should apply RC calibration
+		bool shouldApplyCalibration = true;
+		
+		// If we are in failsafe mode - don't bother
```
Comment (ledvinap) [inline src/main/rx/rx.c]: Is there any reason for this?

**Comment by digitalentity** [comment]
There probably might be more than one reason to avoid using calibration. Failsafe mode is one of cases when firmware sets a predefined channel value, we should not mess with that value.
I am not very familiar with cleanflight architecture yet, and probably missed other cases we shouldn't use calibration. I decided to make adding new cases simple - just add another if() that resets the shouldApplyCalibration variable.
Currently the 'shouldApplyCalibration' variable can easily be removed, if not needed.

**Comment by thenickdude** [comment]
Hm, isn't this a problem then? Because won't the calibration will be
applied *before* the failsafe gets to look at the PWM values? So it may
prevent the quad from entering failsafe mode by messing with the values by
calibration, and the shouldApplyCalibration will never fire (since failsafe
won't activate).​

**Comment by digitalentity** [comment]
@thenickdude, thanks!!! I now see how we may enter failsafe mode, but never exit it.

Imagine a transmitter with throttle range 500 .. 900. If we feed this directly into failsafe, we will always end up in failsafe mode regardless of throttle stick. However, if we apply calibration and map this range to 1000..2000 we will have a sane throttle value. Raw uncalibrated values outside the 500..900 range (a bug, a transmitter failure, whatever) will map outside the 1000..2000 range and thus trigger a failsafe. No PWM or PPM signal equals to raw channel value of 0 (zero), this will also map outside of an expected 1000..2000 range and also trigger a failsafe.
*But* if we don't apply calibration in failsafe mode, we'll never see a valid channel value again and will never exit failsafe mode. 
I've removed this check all together. The new code will apply calibration to any received data.

**Comment by thenickdude** [comment]
That seems tricky to me, since now if you happen to know that your receiver outputs 800us on failure (say), setting the failsafe threshold to 850 has no relationship to the 800us you know the receiver is sending, because the receiver value will be scaled by calibration before the failsafe threshold of 850us is applied. 

You would however be able to set the correct threshold by looking at what the Rx sends on the receiver tab, since those have been scaled already.

**Comment by digitalentity** [comment]
And that's exactly why calibration should be applied in all cases and before failsafe checks.

**Comment by thenickdude** [comment]
You also need to account for scaling causing underflow. For example consider the case when minrc = 1500, maxrc = 2000, and your receiver sends 900us on failsafe, which seems like a reasonable situation to me:

```c
#include <stdint.h>
#include <stdio.h>

int scaleRange(int x, int srcMin, int srcMax, int destMin, int destMax) {
    long int a = ((long int) destMax - (long int) destMin) * ((long int) x - (long int) srcMin);
    long int b = (long int) srcMax - (long int) srcMin;
    return ((a / b) - (destMax - destMin)) + destMax;
}

int main(void)
{
    uint16_t sample, scaledSample;

    // Say that rx failsafe sends 900us
    sample = 900;

    scaledSample = scaleRange(sample, 1500, 2000, 1000, 2000);

    printf("Input %u -> after scaling %u\n", sample, scaledSample);

    return 0;
}
```

Prints:

```
Input 900 -> after scaling 65336
```

**Comment by thenickdude** [comment]
Since values outside the range 1000...2000 get replaced by midrc it seems
like you can never find out what us your receiver sends on fail-safe when
it's outside the normal stick range? It won't appear on the receiver tab as
far as I can tell, so you can't tell what to set your fail-safe threshold
to.

**Comment by digitalentity** [comment]
I'll look into it, thanks for pointing this out.

**Comment by digitalentity** [comment]
I think the issue with failsafe mode is now fixed. Actual stick ranges will still scale to [1000;2000] range,  slightly out of range values will be scaled to [750;2250] range. Significantly out of range input values will be mapped to 750 or 2250 (within valid pulse range but outside of normal transmitter range, thus allowing failsafe to trigger). A special case with no PPM/PWM signal is also handled corretly, triggers failsafe and gets replaced by midrc value.

**Comment by digitalentity** [comment]
Tested this today on actual CC3D board. Works fine, does not prevent failsafe from triggering on PPM/PWM signal loss and pulse out of range. Should be safe to merge now.

**Comment by digitalentity** [comment]
Flew my quadcopter yesterday with this patch applied. Works fine, failsafe triggers as it should. Did a rebase to most recent 'master' so it should merge cleanly now.

**Comment by digitalentity** [comment]
I am closing this as I am reconfiguring my forked repo. I'll reopen a PR after that.

**Comment by digitalentity** [comment]
A new PR #618 created. Also slightly re-worked the code.

**Pull request closed without merging**